File: leetcode/algorithms/stack/reverse-substrings-between-each-pair-of-parentheses.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Substrings Between Each Pair of Parentheses"
*/

class Solution {
    public String reverseParentheses(String s) {
        StringBuilder resultBuilder = new StringBuilder("");
        Stack<Character> charStack = new Stack<Character>();
        
        LinkedList<Character> charQueue = new LinkedList<Character>();
        
        for(int i=0;i<s.length();++i){
            if(s.charAt(i)==')'){
                while(charStack.peek()!='('){
                    charQueue.add(charStack.pop());
                }
                charStack.pop();
                while(charQueue.size()>0){
                    charStack.push(charQueue.poll());
                }
            } else {
                charStack.push(s.charAt(i));
            }
        }
        
        while(!charStack.empty()){
            resultBuilder.append(charStack.pop());
        }
        
        return resultBuilder.reverse().toString();
         
        
    }
}

File: leetcode/algorithms/stack/reverse-substrings-between-each-pair-of-parentheses.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Reverse Substrings Between Each Pair of Parentheses"
*/

/**
 * @param {string} s
 * @return {string}
 */

const _private_stack_map = new WeakMap();

const Stack = function(){
	const _stackData = {
		stack: []
	}
	_private_stack_map.set(this,_stackData);
	
}


Stack.prototype.push = function(val){
	let stackData = _private_stack_map.get(this);
	stackData.stack.push(val);
}

Stack.prototype.pop = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.pop();
}

Stack.prototype.peek = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.length===0 ? null : stackData.stack[stackData.stack.length-1];
}

Stack.prototype.getSize = function(){
	let stackData = _private_stack_map.get(this);
	return stackData.stack.length;
}

var reverseParentheses = function(s) {
    const charStack = new Stack();
    const charBuff = [];
    
    for(let i=0;i<s.length;++i){
        if(s.charAt(i)===')'){
            while(charStack.peek()!='('){
                charBuff.push(charStack.pop());
            }
            charStack.pop();
            while(charBuff.length>0){
                charStack.push(charBuff.shift());
            }
        } else {
            charStack.push(s.charAt(i));
        }
    }
    
    const strArr = Array.from({length: charStack.getSize()});
    let i=charStack.getSize();
    while(charStack.peek()){
        strArr[i] = charStack.pop();
        --i;
    }
    
    return strArr.join('');
    
    
};